src/Ouroboros/Network/FetchLogic.hs

```diff
@@ -0,0 +1,1258 @@
+and we have the task of selecting which block bodies to download, and then
+passing those block bodes back to the consensus layer. The consensus layer will
+try to validate them and decide if it wants to update its current chain.
+
+    ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆
+    ├───┤     ├───┤     ├───┤     ├───┤     ├───┤
+    │   │     │   │     │   │     │   │     │   │
+    ├───┤     ├───┤     ├───┤     ├───┤     ├───┤
+    │   │     │   │     │   │     │   │     │   │
+    ├───┤     ├───┤     ├───┤     ├───┤     ├───┤
+    │   │     │   │     │   │     │   │     │   │
+    ├───┤     ├───┤     ├───┤     ├───┤     └───┘
+    │   │     │   │     │   │     │   │
+ ───┴───┴─────┼───┼─────┼───┼─────┼───┼───────────── current chain length
+              │   │     │   │     │   │
+  current     ├───┤     ├───┤     └───┘
+  (blocks)    │   │     │   │
+              └───┘     └───┘
+                A         B         C         D
+             candidates
+             (headers)
+
+In this example we have four candidate chains, with all but chain D strictly
+longer than our current chain.
+
+In general there are many candidate chains. We make a distinction between a
+candidate chain and the peer from which it is available. It is often the
+case that the same chain is available from multiple peers. We will try to be
+clear about when we are referring to chains or the combination of a chain and
+the peer from which it is available.
+
+For the sake of the example let us assume we have the four chains above
+available from the following peers.
+
+peer    1         2         3         4         5         6         7
+      ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆     ┆   ┆
+      ├───┤     ├───┤     ├───┤     ├───┤     ├───┤     ├───┤     ├───┤
+      │   │     │   │     │   │     │   │     │   │     │   │     │   │
+      ├───┤     ├───┤     ├───┤     ├───┤     └───┘     ├───┤     ├───┤
+      │   │     │   │     │   │     │   │               │   │     │   │
+    ──┼───┼─────┼───┼─────┼───┼─────┼───┼───────────────┼───┼─────┼───┼──
+      │   │     │   │     │   │     │   │               │   │     │   │
+      └───┘     ├───┤     ├───┤     ├───┤               ├───┤     ├───┤
+                │   │     │   │     │   │               │   │     │   │
+                └───┘     └───┘     └───┘               └───┘     └───┘
+chain   C         A         B         A         D         B         A
+
+This is the form in which we are informed about candidate chains from the
+consensus layer, the combination of a chain and the peer it is from. This
+makes sense, since these things change independently.
+
+We will process the chains in this form, keeping the peer/chain combination all
+the way through. Although there could in principle be some opportunistic saving
+by sharing when multiple peers provide the same chain, taking advantage of this
+adds complexity and does nothing to improve our worst case costs.
```

Comment:
Is this because the worst case is fetching long chains from different peers? What if the network was partitioned in two and our node is one of the nodes in the middle.  I think it would make it more efficient in this case.